{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "82028a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import joblib\n",
    "import argparse\n",
    "from sklearn.ensemble import RandomForestClassifier  # random forest model\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn import metrics  # to calculate the accuracy of the model\n",
    "from sklearn.metrics import (precision_recall_curve, PrecisionRecallDisplay, auc)\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.inspection import permutation_importance\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio import SeqIO\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "import random\n",
    "from sklearn.ensemble import RandomForestClassifier  # random forest model\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn import metrics  # to calculate the accuracy of the model and confusion matrix\n",
    "from sklearn.metrics import (precision_recall_curve, PrecisionRecallDisplay, auc)\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.inspection import permutation_importance\n",
    "from sklearn import tree\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio import SeqIO\n",
    "from imblearn.under_sampling import RandomUnderSampler\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9eb9a66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD THE MODEL\n",
    "common_dir = \"/home/giulia/Workspace/PhytoPhD/period_abroad/\"\n",
    "# eff_predictions = f\"{common_dir}random_forest/tmp/old_validation_pre_nizza/feature_table_TNS.tsv\"\n",
    "eff_predictions = f\"{common_dir}random_forest/on_dataset20230215/pred_validation/feature_table_eff25_58_val.tsv\"\n",
    "non_eff_predictions = f\"{common_dir}random_forest/on_dataset20230215/pred_validation/feature_table_noneff_val.tsv\"\n",
    "loaded_model = joblib.load(f\"{common_dir}random_forest/on_dataset20230214/rf_training/rf_best_model.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e665ab5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = [\"name\", \"sequence length\", \"signal peptide\", \"transmembrane domain\", \n",
    "                                 \"aa in tr domain\", \"first 60 aa\", \"prob N-in\", \"warning signal sequence\", \n",
    "                                 \"MobiDB-lite\", \"CLUMP0\", \"CLUMP1\", \"CLUMP5\", \n",
    "                                 \"CLUMP6\", \"CLUMP7\", \"CLUMP8\",\"bin1\", \"bin2\", \"bin3\", \"bin4\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "d136d9b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FOR SINGLE CLASS USAGE \n",
    "class_name = \"effectors\"\n",
    "class_file = \"/home/giulia/Workspace/PhytoPhD/effectors_analysis/classification/prediction_tools/pred_dataset_validazione20221110/feature_table_eff/feature_table_eff_truly_never_seen20221115.tsv\"\n",
    "\n",
    "data_pred_df_original = pd.read_csv(class_file, sep=\"\\t\", header=0)\n",
    "data_pred_df = pd.read_csv(class_file, sep=\"\\t\", header=0)\n",
    "data_pred_df[\"name\"] = [class_name] * len(data_pred_df)\n",
    "data_pred_df.replace([True, False], [\"1\", \"0\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebf8a3af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PUTATIVE EFFECTORS EXPLICITED\n",
    "eff_pred_df_original = pd.read_csv(eff_predictions, sep=\"\\t\", header=0)\n",
    "non_eff_pred_df_original = pd.read_csv(non_eff_predictions, sep=\"\\t\", header=0)\n",
    "data_pred_df_original = pd.concat([eff_pred_df_original, non_eff_pred_df_original], ignore_index=True, join=\"inner\")\n",
    "data_pred_df_original.replace([True, False], [\"1\", \"0\"], inplace=True)\n",
    "\n",
    "print(data_pred_df_original)\n",
    "ids = []\n",
    "\n",
    "with open(\"/home/giulia/Workspace/PhytoPhD/effectors_analysis/dataset_validazione20221110/uniprot-download_true_format_list_query__28Putative_29-2022.11.15-15.57.14.63.list\", \n",
    "          \"r\") as only_putative_ids: \n",
    "    asd = only_putative_ids.readlines()\n",
    "    for el in asd:\n",
    "        ids.append(el.replace(\"\\n\", \"\"))\n",
    "    print(ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "9cb472d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n",
      "             name  sequence length  signal peptide  transmembrane domain  \\\n",
      "0        effector              112           0.886                     1   \n",
      "1        effector              112           0.886                     1   \n",
      "2        effector              112           0.886                     1   \n",
      "3        effector              112           0.886                     1   \n",
      "4        effector              112           0.886                     1   \n",
      "..            ...              ...             ...                   ...   \n",
      "523  non_effector              732           0.628                     1   \n",
      "524  non_effector              357           0.000                     1   \n",
      "525  non_effector              145           0.691                     1   \n",
      "526  non_effector              135           0.000                     1   \n",
      "527  non_effector              635           0.648                     1   \n",
      "\n",
      "     aa in tr domain  first 60 aa  prob N-in  warning signal sequence  \\\n",
      "0           22.14889     22.14889    0.85374                        1   \n",
      "1           22.14889     22.14889    0.85374                        1   \n",
      "2           22.14889     22.14889    0.85374                        1   \n",
      "3           22.14889     22.14889    0.85374                        1   \n",
      "4           22.14889     22.14889    0.85374                        1   \n",
      "..               ...          ...        ...                      ...   \n",
      "523         20.23322     20.23322    0.99978                        1   \n",
      "524         21.92103     21.78263    0.79670                        1   \n",
      "525         21.32152     21.31520    0.79407                        1   \n",
      "526         20.39441     20.39107    0.83093                        1   \n",
      "527         22.56104     22.54575    0.99633                        1   \n",
      "\n",
      "     MobiDB-lite  CLUMP1  CLUMP3  CLUMP4  CLUMP7  CLUMP8  CLUMP9  bin1  bin2  \\\n",
      "0           20.0       0       3       0       0       1       1     1     1   \n",
      "1           20.0       0       3       0       0       1       1     1     1   \n",
      "2           20.0       0       3       0       0       1       1     1     1   \n",
      "3           20.0       0       3       0       0       1       1     1     1   \n",
      "4           20.0       0       3       0       0       1       1     1     1   \n",
      "..           ...     ...     ...     ...     ...     ...     ...   ...   ...   \n",
      "523          0.0       0       1       2       1       1       2     3     4   \n",
      "524          0.0       0       1       0       0       0       1     2     0   \n",
      "525          0.0       0       1       0       0       0       0     0     0   \n",
      "526          0.0       0       1       1       0       0       1     1     1   \n",
      "527          0.0       1       2       2       0       1       2     2     3   \n",
      "\n",
      "     bin3  bin4  \n",
      "0       1     2  \n",
      "1       1     2  \n",
      "2       1     2  \n",
      "3       1     2  \n",
      "4       1     2  \n",
      "..    ...   ...  \n",
      "523     2     4  \n",
      "524     0     1  \n",
      "525     0     1  \n",
      "526     0     1  \n",
      "527     2     1  \n",
      "\n",
      "[528 rows x 19 columns]\n"
     ]
    }
   ],
   "source": [
    "# CONCAT EFF non-EFF VALIDATION SET \n",
    "\n",
    "original_protein_names = pd.concat([pd.read_csv(eff_predictions, sep=\"\\t\", header=0)[[\"name\"]], \n",
    "                                    pd.read_csv(non_eff_predictions, sep=\"\\t\", header=0)[[\"name\"]]], \n",
    "                                   ignore_index=True, join=\"inner\")\n",
    "\n",
    "## EFF PREDICTIONS DF for Random Forest\n",
    "eff_pred_df = pd.read_csv(eff_predictions, sep=\"\\t\", header=0)\n",
    "# eff_pred_df[\"name\"] = [\"effector\"] * len(eff_pred_df)\n",
    "\n",
    "## NON-EFF PREDICTIONS DF for Random Forest\n",
    "non_eff_pred_df = pd.read_csv(non_eff_predictions, sep=\"\\t\", header=0)\n",
    "# non_eff_pred_df[\"name\"] = [\"non_effector\"] * len(non_eff_pred_df)\n",
    "\n",
    "## COMBINE DFs (taking only effector motifs)\n",
    "for col in eff_pred_df.columns[1:]:\n",
    "    if col not in non_eff_pred_df.columns[1:]:\n",
    "        # if an effector motifs is not present in non effectors, a new column will be added to the non_effector_df\n",
    "        # having value False or 0 in this case\n",
    "        non_eff_pred_df[col] = [0] * len(non_eff_pred_df)\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "for feature in features:\n",
    "    if feature not in list(eff_pred_df.columns):\n",
    "        eff_pred_df[feature] = [0] * len(eff_pred_df)\n",
    "    if feature not in list(non_eff_pred_df.columns):\n",
    "        non_eff_pred_df[feature] = [0] * len(non_eff_pred_df)\n",
    "        \n",
    "# CONCAT AND CLEAN UP THE DATAFRAME FOR THE TRAINING\n",
    "# \"inner\" will take only common cols\n",
    "data_pred_df = pd.concat([eff_pred_df, non_eff_pred_df], ignore_index=True, join=\"inner\")\n",
    "data_pred_df.replace([True, False], [\"1\", \"0\"], inplace=True)\n",
    "data_pred_df = data_pred_df[data_pred_df.columns[1:]]\n",
    "\n",
    "print(len(features))\n",
    "print(data_pred_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "773f9200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PARSE THE REAL DATA\n",
    "\n",
    "# data_pred_df = pd.read_csv(data_predictions, sep=\"\\t\", header=0)\n",
    "# data_pred_df.replace([True, False], [\"1\", \"0\"], inplace=True)\n",
    "\n",
    "cols_to_drop = []\n",
    "\n",
    "for col in list(data_pred_df.columns):\n",
    "    if col not in features:\n",
    "        cols_to_drop.append(col)\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "data_pred_df = data_pred_df.drop(columns=cols_to_drop, axis=1)\n",
    "\n",
    "for col in features:\n",
    "    if col not in data_pred_df.columns:\n",
    "        data_pred_df[col] = [0] * len(data_pred_df)\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "data_pred_df = data_pred_df[features]\n",
    "len(data_pred_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67557ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PROB ON PREDICITONS\n",
    "print(loaded_model.feature_names_in_)\n",
    "y_proba = loaded_model.predict_proba(data_pred_df[list(data_pred_df.columns)[1:]])\n",
    "pred_proba_df = pd.DataFrame(y_proba, columns=loaded_model.classes_)\n",
    "\n",
    "pred_proba_df[\"name\"] = original_protein_names\n",
    "pred_proba_df[\"known_clf\"] = data_pred_df[\"name\"]\n",
    "pred_proba_df.to_csv(f\"{common_dir}random_forest/rf_clumps_dataset_validazione20230131/predictions_probability_on_effTNS_negrandom100_20221116_20230131.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index=True)\n",
    "\n",
    "pred_proba_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dee6057a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     sequence length  signal peptide  transmembrane domain  aa in tr domain  \\\n",
      "0                741           0.000                     0          0.00000   \n",
      "1                394           0.000                     0          0.22113   \n",
      "2                211           0.000                     0          0.04486   \n",
      "3                501           0.678                     1         20.64188   \n",
      "4                263           0.700                     1         20.77388   \n",
      "..               ...             ...                   ...              ...   \n",
      "514              233           0.000                     5        109.90518   \n",
      "515              286           0.735                     1         20.30865   \n",
      "516              354           0.000                     0          0.00422   \n",
      "517               54           0.000                     0          0.00018   \n",
      "518              329           0.000                     0          0.00237   \n",
      "\n",
      "     first 60 aa  prob N-in  warning signal sequence  MobiDB-lite  CLUMP0  \\\n",
      "0        0.00000    0.00007                        0          0.0       1   \n",
      "1        0.00071    0.06997                        0          0.0       0   \n",
      "2        0.00018    0.24045                        0          0.0       0   \n",
      "3       20.63883    0.99701                        1          0.0       1   \n",
      "4       20.76635    0.98330                        1          0.0       2   \n",
      "..           ...        ...                      ...          ...     ...   \n",
      "514     34.96174    0.95920                        1          0.0       0   \n",
      "515     20.30697    0.96615                        1          0.0       1   \n",
      "516      0.00098    0.01808                        0          0.0       0   \n",
      "517      0.00018    0.28323                        0          0.0       0   \n",
      "518      0.00000    0.02354                        0          0.0       1   \n",
      "\n",
      "     CLUMP3  CLUMP5  CLUMP6  CLUMP7  CLUMP8  bin1  bin2  bin3  bin4  \n",
      "0         1       0       2       5       1    14    16     3     5  \n",
      "1         0       0       0       0       0     0     0     0     0  \n",
      "2         0       0       0       0       0     0     0     0     0  \n",
      "3         0       0       3       1       0     0     3     2     0  \n",
      "4         0       1       0       2       1     3     2     1     0  \n",
      "..      ...     ...     ...     ...     ...   ...   ...   ...   ...  \n",
      "514       0       0       0       1       0     0     0     1     0  \n",
      "515       0       0       1       1       0     0     0     1     2  \n",
      "516       0       0       0       3       1     0     0     1     3  \n",
      "517       0       0       0       0       0     0     0     0     0  \n",
      "518       0       2       1       1       1     1     3     1     1  \n",
      "\n",
      "[519 rows x 18 columns]\n",
      "accuracy: 0.0\n",
      "conf_matrix:\n",
      "[[  0   0   0]\n",
      " [  0   0   0]\n",
      " [117 402   0]]\n",
      "precision: 1.0\n",
      "recall: 0.2254335260115607\n"
     ]
    }
   ],
   "source": [
    "# APPLY TO REAL DATA\n",
    "\n",
    "# data_pred_df_putative_exp_col = []\n",
    "# ids = []\n",
    "# for i in range(len(data_pred_df_original)):\n",
    "#     if data_pred_df_original[\"name\"].iloc[i] in ids:\n",
    "#         data_pred_df_putative_exp_col.append(\"putative_effectors\")\n",
    "#     else:\n",
    "#         data_pred_df_putative_exp_col.append(\"effectors\")\n",
    "loaded_model = joblib.load(f\"{base_dir}on_dataset20230224/clumps_features/rf_best_model.pkl\")\n",
    "clf = loaded_model.predict(data_pred_df[list(data_pred_df.columns)[1:]])\n",
    "list_clf = list(clf)\n",
    "\n",
    "# clf_df = pd.DataFrame(list(zip(list(original_protein_names), list(data_pred_df[\"name\"]), clf)),\n",
    "#                               columns=[\"uniprot_name\", \"class\", \"RF_model_classification\"])\n",
    "\n",
    "# print(clf_df[\"RF_model_classification\"].value_counts())\n",
    "# clf_df.to_csv(f\"{common_dir}random_forest/rf_clumps_dataset_validazione20230213/clf_on_effTNS_negrandom100fake_20221116.tsv\",\n",
    "#              sep=\"\\t\",\n",
    "#              index=False)\n",
    "# print(data_pred_df[\"name\"], clf)\n",
    "acc = metrics.accuracy_score(data_pred_df[\"name\"], clf)\n",
    "conf_matrix = metrics.confusion_matrix(data_pred_df[\"name\"], clf)\n",
    "known_hot_encode = [0 if el == \"non_effector\" else 1 for el in data_pred_df[\"name\"]]\n",
    "clf_hot_encode = [0 if el == \"non_effector\" else 1 for el in clf]\n",
    "precision_perc = metrics.precision_score(known_hot_encode, clf_hot_encode, pos_label=1) # calculated on positive class\n",
    "recall_perc = metrics.recall_score(known_hot_encode, clf_hot_encode, pos_label=1) # calculated on positive class\n",
    "print(f\"accuracy: {acc}\")\n",
    "print(f\"conf_matrix:\\n{conf_matrix}\")\n",
    "print(f\"precision: {precision_perc}\")\n",
    "print(f\"recall: {recall_perc}\")\n",
    "pd.DataFrame({\"clf\": list_clf, \"real\": list(data_pred_df[\"name\"]), \"seq_id\": list(origin_data_pred_df[\"seq_id\"])}).to_csv(f\"{base_dir}on_uniproteome_mali20230214/rf_validation/protein_classification.tsv\")\n",
    "# with open(f\"{common_dir}random_forest/on_dataset20230214/rf_validation/summary_performances.txt\", \"w\") as out:\n",
    "# #     out.write(f\"{list(clf_df['RF_model_classification'].value_counts().index)[0]}\\t{clf_df['RF_model_classification'].value_counts()[0]}\\n\")\n",
    "# #     out.write(f\"{list(clf_df['RF_model_classification'].value_counts().index)[1]}\\t{clf_df['RF_model_classification'].value_counts()[1]}\\n\")\n",
    "#     out.write(f\"accuracy\\t{acc}\\nconf_matrix:\\n{conf_matrix}\\nprecision\\t{precision_perc}\\nrecall\\t{recall_perc}\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "691d0698",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sequence length</th>\n",
       "      <th>signal peptide</th>\n",
       "      <th>transmembrane domain</th>\n",
       "      <th>aa in tr domain</th>\n",
       "      <th>first 60 aa</th>\n",
       "      <th>prob N-in</th>\n",
       "      <th>warning signal sequence</th>\n",
       "      <th>MobiDB-lite</th>\n",
       "      <th>CAMP_PHOSPHO_SITE</th>\n",
       "      <th>...</th>\n",
       "      <th>CLUMP0</th>\n",
       "      <th>CLUMP3</th>\n",
       "      <th>CLUMP5</th>\n",
       "      <th>CLUMP6</th>\n",
       "      <th>CLUMP7</th>\n",
       "      <th>CLUMP8</th>\n",
       "      <th>bin1</th>\n",
       "      <th>bin2</th>\n",
       "      <th>bin3</th>\n",
       "      <th>bin4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>356</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1</td>\n",
       "      <td>21.91817</td>\n",
       "      <td>21.77976</td>\n",
       "      <td>0.79473</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>842</td>\n",
       "      <td>0.628</td>\n",
       "      <td>1</td>\n",
       "      <td>20.23306</td>\n",
       "      <td>20.23306</td>\n",
       "      <td>0.99993</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>67</td>\n",
       "      <td>0.574</td>\n",
       "      <td>1</td>\n",
       "      <td>21.91432</td>\n",
       "      <td>21.90733</td>\n",
       "      <td>0.03253</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>132</td>\n",
       "      <td>0.736</td>\n",
       "      <td>0</td>\n",
       "      <td>10.61466</td>\n",
       "      <td>10.61226</td>\n",
       "      <td>0.41875</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>601</td>\n",
       "      <td>0.637</td>\n",
       "      <td>2</td>\n",
       "      <td>45.16399</td>\n",
       "      <td>22.44981</td>\n",
       "      <td>0.00499</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>541</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>75</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.00056</td>\n",
       "      <td>0.21904</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>267</td>\n",
       "      <td>0.000</td>\n",
       "      <td>7</td>\n",
       "      <td>154.50996</td>\n",
       "      <td>22.62559</td>\n",
       "      <td>0.98469</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>232</td>\n",
       "      <td>0.807</td>\n",
       "      <td>1</td>\n",
       "      <td>34.07659</td>\n",
       "      <td>12.43419</td>\n",
       "      <td>0.60223</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>30</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.80363</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>to_be_predicted</td>\n",
       "      <td>55</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00036</td>\n",
       "      <td>0.00036</td>\n",
       "      <td>0.52825</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>546 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                name  sequence length  signal peptide  transmembrane domain  \\\n",
       "0    to_be_predicted              356           0.000                     1   \n",
       "1    to_be_predicted              842           0.628                     1   \n",
       "2    to_be_predicted               67           0.574                     1   \n",
       "3    to_be_predicted              132           0.736                     0   \n",
       "4    to_be_predicted              601           0.637                     2   \n",
       "..               ...              ...             ...                   ...   \n",
       "541  to_be_predicted               75           0.000                     0   \n",
       "542  to_be_predicted              267           0.000                     7   \n",
       "543  to_be_predicted              232           0.807                     1   \n",
       "544  to_be_predicted               30           0.000                     0   \n",
       "545  to_be_predicted               55           0.000                     0   \n",
       "\n",
       "     aa in tr domain  first 60 aa  prob N-in  warning signal sequence  \\\n",
       "0           21.91817     21.77976    0.79473                        1   \n",
       "1           20.23306     20.23306    0.99993                        1   \n",
       "2           21.91432     21.90733    0.03253                        1   \n",
       "3           10.61466     10.61226    0.41875                        1   \n",
       "4           45.16399     22.44981    0.00499                        1   \n",
       "..               ...          ...        ...                      ...   \n",
       "541          0.00056      0.00056    0.21904                        0   \n",
       "542        154.50996     22.62559    0.98469                        1   \n",
       "543         34.07659     12.43419    0.60223                        1   \n",
       "544          0.00000      0.00000    0.80363                        0   \n",
       "545          0.00036      0.00036    0.52825                        0   \n",
       "\n",
       "     MobiDB-lite  CAMP_PHOSPHO_SITE  ...  CLUMP0  CLUMP3  CLUMP5  CLUMP6  \\\n",
       "0              0                  0  ...       1       0       1       0   \n",
       "1              0                  0  ...       2       1       0       2   \n",
       "2              0                  0  ...       0       0       0       0   \n",
       "3             23                  0  ...       0       0       0       0   \n",
       "4              0                  0  ...       1       0       1       1   \n",
       "..           ...                ...  ...     ...     ...     ...     ...   \n",
       "541            0                  1  ...       0       0       0       0   \n",
       "542            0                  0  ...       0       0       0       0   \n",
       "543            0                  0  ...       0       0       0       0   \n",
       "544            0                  0  ...       0       0       0       0   \n",
       "545            0                  0  ...       0       0       0       0   \n",
       "\n",
       "     CLUMP7  CLUMP8  bin1  bin2  bin3  bin4  \n",
       "0         3       0     5     0     0     1  \n",
       "1         6       1     3     5     2    10  \n",
       "2         0       0     0     0     0     0  \n",
       "3         2       0     0     1     0     1  \n",
       "4         3       1     2     1     0     4  \n",
       "..      ...     ...   ...   ...   ...   ...  \n",
       "541       1       0     0     0     1     0  \n",
       "542       2       0     2     0     1     0  \n",
       "543       2       0     1     0     0     2  \n",
       "544       0       0     0     0     0     0  \n",
       "545       0       0     0     0     0     0  \n",
       "\n",
       "[546 rows x 31 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## unique feature table \n",
    "base_dir = \"/home/giulia/Workspace/PhytoPhD/period_abroad/random_forest/\"\n",
    "origin_data_pred_df = pd.read_csv(f\"{base_dir}on_7_strains/capm40/pred/feature_table_capm40_clumps_std.tsv\",\n",
    "                          sep=\"\\t\", header=0)\n",
    "original_protein_names = origin_data_pred_df[\"name\"]\n",
    "\n",
    "data_pred_df = origin_data_pred_df[origin_data_pred_df.columns[1:]] # df without seq_id \n",
    "data_pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "02cd93ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "74\n",
      "0.13553113553113552\n",
      "Predicted 193 effectors and 353 non-effectors\n",
      "Proportion of effectors/total proteins: 54.6742209631728%\n"
     ]
    }
   ],
   "source": [
    "## APPLY AFTER CROSS VALIDATION \n",
    "# basic names for application\n",
    "model_feats = \"clumps_std\"\n",
    "bn_app = \"capm40\"\n",
    "bn_prot = \"CaPm40\"\n",
    "\n",
    "# load and apply the model\n",
    "loaded_model = joblib.load(f\"{base_dir}on_dataset20230224/{model_feats}_features/rf_best_model.pkl\")\n",
    "clf = loaded_model.predict(data_pred_df[list(data_pred_df.columns)[1:]])\n",
    "list_clf = list(clf)\n",
    "\n",
    "# probability of class assignment \n",
    "prob = loaded_model.predict_proba(data_pred_df[list(data_pred_df.columns)[1:]])\n",
    "df_prob = pd.DataFrame(prob, columns=[\"effector_prob\", \"non_effector_prob\"])\n",
    "df_prob[\"seq_id\"] = list(origin_data_pred_df[\"seq_id\"])\n",
    "df_prob.to_csv(f\"{base_dir}on_7_strains/{bn_app}/model_{model_feats}_on_dataset20230224/probability_of_predictions.tsv\", sep=\"\\t\", index=False)\n",
    "print(len(df_prob[df_prob[\"effector_prob\"] > 0.9]))\n",
    "print(len(df_prob[df_prob[\"effector_prob\"] > 0.9])/len(origin_data_pred_df))\n",
    "highly_eff = df_prob[df_prob[\"effector_prob\"] > 0.9]\n",
    "with open(f\"{base_dir}on_7_strains/{bn_app}/model_{model_feats}_on_dataset20230224/effector_predictions_filt90.fasta\", \"w\") as filtered_seq:\n",
    "    for record in SeqIO.parse(f\"/home/giulia/Workspace/PhytoPhD/period_abroad/datasets/7_strains/{bn_prot}_prodigal_faa.fasta\", \"fasta\"):\n",
    "        if record.id in list(highly_eff[\"seq_id\"]):\n",
    "            filtered_seq.write(f\">{record.description}\\n{str(record.seq)}\\n\")\n",
    "\n",
    "# save the summary of performances \n",
    "with open(f\"{base_dir}on_7_strains/{bn_app}/model_{model_feats}_on_dataset20230224/prediction_summary.txt\", \"w\") as out:\n",
    "    count_eff = list_clf.count('effector')\n",
    "    count_noneff = list_clf.count('non_effector')\n",
    "    out.write(f\"Predicted {count_eff} effectors and {count_noneff} non-effectors\\nProportion of effectors/total proteins: {(count_eff/count_noneff)*100}\")\n",
    "\n",
    "    \n",
    "# save the predictions as they are\n",
    "df_predictions = pd.DataFrame({\"clf\": list_clf, \"protein_id\": origin_data_pred_df[\"seq_id\"]})\n",
    "df_predictions.to_csv(f\"{base_dir}on_7_strains/{bn_app}/model_{model_feats}_on_dataset20230224/predictions.tsv\", sep=\"\\t\", index=False)\n",
    "    \n",
    "print(f\"Predicted {count_eff} effectors and {count_noneff} non-effectors\\nProportion of effectors/total proteins: {(count_eff/count_noneff)*100}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a5a0173f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "effectors        26\n",
      "non_effectors    12\n",
      "Name: RF_model_classification, dtype: int64\n",
      "accuracy: 0.6842105263157895\n",
      "conf_matrix:\n",
      "[[26 12]\n",
      " [ 0  0]]\n",
      "precision: 1.0\n",
      "recall: 0.6842105263157895\n"
     ]
    }
   ],
   "source": [
    "########################################## OLD #################################################################\n",
    "# APPLY TO known effectors DATA\n",
    "known = [\"effectors\"] * len(pred_data_eff)\n",
    "clf = loaded_model.predict(pred_data_eff[list(pred_data_eff.columns)[1:]])\n",
    "clf_df = pd.DataFrame(list(zip(list(pred_data_eff[\"name\"]), clf)),\n",
    "                              columns=[\"name\", \"RF_model_classification\"])\n",
    "\n",
    "print(clf_df[\"RF_model_classification\"].value_counts())\n",
    "# clf_df.to_csv(\"/home/giulia/Workspace/PhytoPhD/effectors_analysis/classification/random_forest/rf_dataset_validazione20220805/putative_eff_eff_clf.tsv\",\n",
    "#              sep=\"\\t\",\n",
    "#              index=False)\n",
    "\n",
    "\n",
    "acc = metrics.accuracy_score(known, clf)\n",
    "conf_matrix = metrics.confusion_matrix(known, clf)\n",
    "known_hot_encode = [0 if el == \"non_effectors\" else 1 for el in known]\n",
    "clf_hot_encode = [0 if el == \"non_effectors\" else 1 for el in clf]\n",
    "precision_perc = metrics.precision_score(known_hot_encode, clf_hot_encode, pos_label=1)\n",
    "recall_perc = metrics.recall_score(known_hot_encode, clf_hot_encode, pos_label=1)\n",
    "print(f\"accuracy: {acc}\")\n",
    "print(f\"conf_matrix:\\n{conf_matrix}\")\n",
    "print(f\"precision: {precision_perc}\")\n",
    "print(f\"recall: {recall_perc}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657a5ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "clf_fit = loaded_model.fit(pred_data_eff[list(pred_data_eff.columns)[1:]], pred_data_eff[\"name\"])\n",
    "perm_import = permutation_importance(clf_fit, pred_data_eff[list(pred_data_eff.columns)[1:]], pred_data_eff[\"name\"],\n",
    "                                    n_repeats=30, random_state=42, scoring='accuracy')\n",
    "perm_import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80788aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### HYDROPHOBIC PROFILE\n",
    "l = np.array([1, 2, 4, 10, 53, 8, 20, 88])\n",
    "l1 = np.array([5, 7, 8, 1])\n",
    "\n",
    "l_mean = l.mean()\n",
    "l_std = l.std()\n",
    "l1_mean = l1.mean()\n",
    "l1_std = l1.std()\n",
    "\n",
    "# l_1 = (l*l_mean)/l_std\n",
    "# l1_1 = (l1*l1_mean)/l1_std\n",
    "\n",
    "l = ((l-l.min()) / (l.max() - l.min())) / len(l)\n",
    "l1 = ((l1-l1.min()) / (l1.max()-l.min())) / len(l1)\n",
    "\n",
    "# l = l/len(l)\n",
    "# l1 = l1/len(l1)\n",
    "bins = [[l[:5].mean(), l[5:].mean()], [l1[:3].mean(), l1[3:].mean()]]\n",
    "print(l)\n",
    "print(l1)\n",
    "# print(l_1)\n",
    "# print(l1_1)\n",
    "print(bins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0f72f047",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          name  sequence length  signal peptide  aa in tr domain  first 60 aa  \\\n",
      "0   A0A0G7ZN61              302           0.405          0.08054      0.00038   \n",
      "1   A0A1Q1NH89              112           0.886         22.14889     22.14889   \n",
      "2   A0A1Q1NH90              112           0.886         22.14889     22.14889   \n",
      "3   A0A1Q1NH91              112           0.886         22.14889     22.14889   \n",
      "4   A0A1Q1NH92              112           0.886         22.14889     22.14889   \n",
      "5   A0A1Q1NH98              112           0.886         22.14889     22.14889   \n",
      "6   A0A1Q1NH99              112           0.886         22.14889     22.14889   \n",
      "7   A0A1Q1NHA1              112           0.886         22.14889     22.14889   \n",
      "8   A0A1Q1NHA2              112           0.886         22.14889     22.14889   \n",
      "9   A0A1Q1NHA3              112           0.886         22.14889     22.14889   \n",
      "10  A0A1Q1NHA5              112           0.886         22.14889     22.14889   \n",
      "11  A0A1Q1NHA6              112           0.886         22.14889     22.14889   \n",
      "12  A0A1Q1NHA7              112           0.818         21.54163     21.54163   \n",
      "13  A0A1S2NIQ0               99           0.875         19.19138     19.19137   \n",
      "14  A0A1S9LYK0              137           0.872         23.45334     23.14639   \n",
      "15  A0A1S9LZ16              130           0.748         22.38050     22.37936   \n",
      "16  A0A1S9M1J1              117           0.654         19.94901     19.94769   \n",
      "17  A0A1S9M2M0              216           0.671         19.60619     19.60619   \n",
      "18  A0A1S9M4Q7               97           0.140          0.00000      0.00000   \n",
      "19  A0A1V0PKQ2              175           0.677         20.90797     20.90734   \n",
      "20  A0A1V0PKR7              115           0.835         21.89528     21.89528   \n",
      "21  A0A1V0PKS9               99           0.798         21.31965     21.31907   \n",
      "22  A0A1V0PKT0              197           0.608         20.27229     20.25730   \n",
      "23  A0A2S8NTF4              158           0.736         20.50741     20.50734   \n",
      "24  A0A2S8NUD4              151           0.611         21.29696     21.29696   \n",
      "25  A0A2S8NUU7              135           0.795         19.62939     19.62779   \n",
      "26  A0A421NUE5              236           0.643         22.61931     22.61858   \n",
      "27  A0A421NUF4              272           0.824         21.34195     21.34154   \n",
      "28  A0A421NUW0              198           0.839         20.62898     20.62898   \n",
      "29  A0A421NV12              195           0.681         18.89826     18.89699   \n",
      "30  A0A421NV20              506           0.729         21.66206     21.66206   \n",
      "31  A0A421NXH2              186           0.523         22.93271     22.72054   \n",
      "32  A0A421NY84              259           0.844         22.79070     22.77504   \n",
      "33  A0A421NYW6              314           0.778         21.40028     21.12889   \n",
      "34  A0A421NZ11              111           0.830         19.77668     19.77668   \n",
      "35  A0A449BEH5              147           0.583         75.34883     34.21805   \n",
      "36  A0A5N5XS43              152           0.638         21.58893     21.58875   \n",
      "37  A0A660HMB1              119           0.633         21.14054     21.14050   \n",
      "38  A0A660HNE9              136           0.803         19.61272     19.58286   \n",
      "39  A0A6M3WSS2              112           0.886         22.14889     22.14889   \n",
      "40  A0A7G1G7V6               44           0.130          0.00000      0.00000   \n",
      "41  A0A7G1G7V7               42           0.104          0.00000      0.00000   \n",
      "42  A0A7G1G7X1              122           0.843         20.55468     20.55468   \n",
      "43  A0A7G1G7X2               43           0.113          0.00000      0.00000   \n",
      "44  A0A7G1G841               43           0.113          0.00000      0.00000   \n",
      "45  A0A7G1G915               43           0.113          0.00000      0.00000   \n",
      "46  A0A7G1GAE5              123           0.775         19.23322     19.23322   \n",
      "47  A0A7G1GBC6               42           0.099          0.00000      0.00000   \n",
      "48  A0A7G1GBD3               42           0.104          0.00000      0.00000   \n",
      "49  A0A7G1GBD4               43           0.113          0.00000      0.00000   \n",
      "50  A0A7G1GCA0               43           0.109          0.00000      0.00000   \n",
      "51  A0A7G1GCL5               43           0.171          0.00000      0.00000   \n",
      "52  A0A7G1GCL6               43           0.113          0.00000      0.00000   \n",
      "53      W8GFX3              303           0.317          0.00097      0.00000   \n",
      "\n",
      "    prob N-in transmembrane_domain  MobiDB-lite  CLUMP0  CLUMP1  CLUMP2  \\\n",
      "0     0.35545                    1            0       2       1       0   \n",
      "1     0.85374                    0           20       7       0       1   \n",
      "2     0.85374                    0           20       7       0       1   \n",
      "3     0.85374                    0           20       7       0       1   \n",
      "4     0.85374                    0           20       7       0       1   \n",
      "5     0.85374                    0           20       7       0       1   \n",
      "6     0.85374                    0           20       7       0       1   \n",
      "7     0.85374                    0           20       7       0       1   \n",
      "8     0.85374                    0           20       7       0       1   \n",
      "9     0.85374                    0           20       7       0       1   \n",
      "10    0.85374                    0           20       7       0       1   \n",
      "11    0.85374                    0           20       7       0       1   \n",
      "12    0.62458                    0           21       5       0       2   \n",
      "13    0.25679                    0           21       1       0       0   \n",
      "14    0.95698                    0            0       0       0       2   \n",
      "15    0.81955                    0            0       0       0       0   \n",
      "16    0.16578                    0            0       0       0       0   \n",
      "17    0.91639                    0            0       0       0       0   \n",
      "18    0.12920                    1           20       0       0       0   \n",
      "19    0.60281                    0           24       1       0       1   \n",
      "20    0.20372                    0            0       1       0       0   \n",
      "21    0.49654                    0           21       1       0       0   \n",
      "22    0.63945                    0            0       0       0       0   \n",
      "23    0.16644                    0            0       0       1       0   \n",
      "24    0.48821                    0            0       0       0       0   \n",
      "25    0.32935                    0            0       0       0       2   \n",
      "26    0.70028                    0            0       1       0       0   \n",
      "27    0.96154                    0            0       0       0       0   \n",
      "28    0.95375                    0           99       2       0       0   \n",
      "29    0.09290                    0            0       1       0       0   \n",
      "30    0.99146                    0           23       1       0       1   \n",
      "31    0.72710                    0            0       0       0       0   \n",
      "32    0.99835                    0            0       0       0       0   \n",
      "33    0.99305                    0            0       1       0       2   \n",
      "34    0.82162                    0           23       1       1       2   \n",
      "35    0.64758                    0            0       0       0       0   \n",
      "36    0.76098                    0            0       4       0       2   \n",
      "37    0.70802                    0            0       1       0       0   \n",
      "38    0.91640                    0            0       1       0       2   \n",
      "39    0.85374                    0           20       7       0       1   \n",
      "40    0.75028                    1            0       0       0       0   \n",
      "41    0.81893                    1            0       0       1       0   \n",
      "42    0.65739                    0            0       0       1       0   \n",
      "43    0.86437                    1            0       1       0       1   \n",
      "44    0.86437                    1            0       1       0       1   \n",
      "45    0.86437                    1            0       1       0       1   \n",
      "46    0.41160                    0            0       1       1       2   \n",
      "47    0.81893                    1            0       0       1       0   \n",
      "48    0.81893                    1            0       0       1       0   \n",
      "49    0.86437                    1            0       1       0       1   \n",
      "50    0.86230                    1            0       1       0       1   \n",
      "51    0.91914                    1            0       1       0       1   \n",
      "52    0.86437                    1            0       1       0       1   \n",
      "53    0.14122                    1            0       1       0       0   \n",
      "\n",
      "    CLUMP3  CLUMP4  CLUMP5  CLUMP6  CLUMP7  CLUMP8  CLUMP9  CLUMP10  \n",
      "0        0       0       0       0       0       0       0        0  \n",
      "1        1       1       2       0       1       0       0        0  \n",
      "2        1       1       2       0       1       0       0        0  \n",
      "3        1       1       2       0       1       0       0        0  \n",
      "4        1       1       2       0       1       0       0        0  \n",
      "5        1       1       2       0       1       0       0        0  \n",
      "6        1       1       2       0       1       0       0        0  \n",
      "7        1       1       2       0       1       0       0        0  \n",
      "8        1       1       2       0       1       0       0        0  \n",
      "9        1       1       2       0       1       0       0        0  \n",
      "10       1       1       2       0       1       0       0        0  \n",
      "11       1       1       2       0       1       0       0        0  \n",
      "12       2       1       3       0       0       0       0        0  \n",
      "13       0       0       0       0       0       0       0        0  \n",
      "14       0       1       0       0       1       0       0        0  \n",
      "15       0       1       0       0       0       1       0        0  \n",
      "16       0       0       1       0       0       0       0        0  \n",
      "17       0       0       1       0       0       0       0        0  \n",
      "18       2       1       0       1       0       1       0        0  \n",
      "19       2       0       0       0       0       0       1        0  \n",
      "20       1       0       0       0       0       0       0        0  \n",
      "21       0       0       2       0       0       0       0        0  \n",
      "22       2       0       1       0       0       0       0        0  \n",
      "23       0       0       0       0       0       0       0        0  \n",
      "24       1       0       0       0       0       0       0        0  \n",
      "25       0       1       0       0       1       0       0        0  \n",
      "26       0       1       0       0       0       0       0        0  \n",
      "27       2       0       1       1       0       0       0        0  \n",
      "28       0       0       2       0       0       0       0        0  \n",
      "29       0       0       0       0       0       1       0        0  \n",
      "30       1       0       3       0       0       0       0        0  \n",
      "31       0       0       0       0       0       1       0        0  \n",
      "32       1       0       0       0       0       0       0        0  \n",
      "33       3       1       0       0       0       0       0        0  \n",
      "34       0       2       2       0       1       0       0        0  \n",
      "35       0       0       1       0       0       0       0        0  \n",
      "36       0       0       0       0       3       0       0        0  \n",
      "37       0       0       0       0       0       0       0        0  \n",
      "38       0       0       2       0       0       1       0        0  \n",
      "39       1       1       2       0       1       0       0        0  \n",
      "40       2       0       0       0       0       0       0        0  \n",
      "41       2       0       0       0       0       0       0        0  \n",
      "42       2       0       1       0       1       1       1        0  \n",
      "43       0       0       1       0       0       0       0        0  \n",
      "44       0       0       1       0       0       0       0        0  \n",
      "45       0       0       1       0       0       0       0        0  \n",
      "46       1       0       0       0       1       1       1        0  \n",
      "47       2       0       0       1       0       0       0        0  \n",
      "48       2       0       0       0       0       0       0        0  \n",
      "49       0       0       1       0       0       0       0        0  \n",
      "50       0       0       1       0       0       0       0        0  \n",
      "51       0       0       1       0       0       0       0        0  \n",
      "52       0       0       1       0       0       0       0        0  \n",
      "53       0       0       0       0       0       0       0        0  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "38"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# effectors \n",
    "# no direct predictions, take the corresponding rows from the putative_eff_eff feature table\n",
    "data_eff = SeqIO.parse(\"/home/giulia/Workspace/PhytoPhD/effectors_analysis/dataset_validazione20220805/uniprot_effectors_never_seen_byRF20220805.fasta\", \"fasta\")\n",
    "name_eff = []\n",
    "for record in data_eff:\n",
    "    name = record.id.split(\"|\")[1]\n",
    "    name_eff.append(name)\n",
    "\n",
    "name_eff_putative_eff = list(data_pred_df[\"name\"])\n",
    "index = []\n",
    "for el in name_eff:\n",
    "    if el in name_eff_putative_eff:\n",
    "        index.append(name_eff_putative_eff.index(el))\n",
    "        \n",
    "pred_data_eff = data_pred_df.iloc[index]\n",
    "\n",
    "len(pred_data_eff)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
